File: Problem147_minesweeper.java
// Time Complexity: O(m * n)
// Space Complexity: O(m * n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    int[][] dirs;
    int m, n;
    public char[][] updateBoard(char[][] board, int[] click) {
        // edge
        if(board == null || board.length == 0) return board;
        dirs = new int[][] {{0, 1}, {0, -1}, {1, 0}, {-1, 0}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};
        this.m = board.length;
        this.n = board[0].length;
        if(board[click[0]][click[1]] == 'M') {
            board[click[0]][click[1]] = 'X';
            return board;
        }
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {click[0], click[1]});
        board[click[0]][click[1]] = 'B';
        while(!q.isEmpty()) {
            int[] curr = q.poll();
            int countmines = countMines(board, curr[0], curr[1]);
            if(countmines > 0) {
                board[curr[0]][curr[1]] = (char)(countmines + '0');
            } else {
                for(int[] dir: dirs) {
                    int r = dir[0] + curr[0];
                    int c = dir[1] + curr[1];
                    if(r >= 0 && c >= 0 && r < m && c < n && board[r][c] == 'E') {
                        q.add(new int[]{r, c});
                        board[r][c] = 'B';
                    }
                }
            }
        }
        return board;
    }
    
    private int countMines(char[][] board, int i, int j) {
        int count = 0;
        for(int[] dir: dirs) {
            int r = dir[0] + i;
            int c = dir[1] + j;
            if(r >= 0 && c >= 0 && r < m && c < n && board[r][c] == 'M') {
                count++;
            }
        }
        return count;
    }
}

File: Problem-147 Minesweeper.py
# 529. Minesweeper
# https://leetcode.com/problems/minesweeper/

# Logic: We start BFS from the click cell given to us. We add all the 
# neighbours if the current mine count is 0 and neighbour is not explored.

# Time Complexity: O(n*m)
# Space Complexity: O(min(n,m))

from collections import deque
class Solution:
    def countBombs(self, board, row, col, dirs):
        n = len(board)
        m = len(board[0])
        count = 0
        
        for _dir in dirs:
            newRow = row + _dir[0]
            newCol = col + _dir[1]

            if 0 <= newRow < n and 0 <= newCol < m and board[newRow][newCol] == 'M':
                count += 1
        return count

    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:
        n = len(board)
        m = len(board[0])

        if board[click[0]][click[1]] == 'M':
            board[click[0]][click[1]] = 'X'
            return board

        q = deque()
        dirs = [(1,0), (-1,0), (0,1), (0,-1), (1,1), (1,-1), (-1,-1), (-1,1)]

        q.append(click)
        board[click[0]][click[1]] = 'B'

        while q:
            row, col = q.popleft()

            mines = self.countBombs(board, row, col, dirs)
            if mines != 0:
                board[row][col] = str(mines)
            else:
                for _dir in dirs:
                    newRow = row + _dir[0]
                    newCol = col + _dir[1]

                    if 0 <= newRow < n and 0 <= newCol < m and board[newRow][newCol] in ['E', 'M']:
                        board[newRow][newCol] = 'B'
                        q.append([newRow, newCol])
            
        return board
